src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -12,100 +12,473 @@
+  } else {
+    // This is the first item in the queue.
+    update.next = queue.first;
+    queue.first = update;
+  }
+
+  if (insertBefore) {
+    update.next = insertBefore;
+  } else {
+    // This is the last item in the queue.
+    queue.last = update;
+  }
+}
+
+// The work-in-progress queue is a subset of the current queue (if it exists).
+// We need to insert the incoming update into both lists. However, it's possible
+// that the correct position in one list will be different from the position in
+// the other. Consider the following case:
+//
+//     Current:             3-5-6
+//     Work-in-progress:        6
+//
+// Then we receive an update with priority 4 and insert it into each list:
+//
+//     Current:             3-4-5-6
+//     Work-in-progress:        4-6
+//
+// In the current queue, the new update's `next` pointer points to the update
+// with priority 5. But in the work-in-progress queue, the pointer points to the
+// update with priority 6. Because these two queues share the same persistent
+// data structure, this won't do. (This can only happen when the incoming update
+// has higher priority than all the updates in the work-in-progress queue.)
+//
+// To solve this, in the case where the incoming update needs to be inserted
+// into two different positions, we'll make a clone of the update and insert
+// each copy into a separate queue. This forks the list while maintaining a
+// persistent stucture, because the update that is added to the work-in-progress
+// is always added to the front of the list.
+//
+// However, if incoming update is inserted into the same position of both lists,
+// we shouldn't make a copy.
+
+function insertUpdate(fiber : Fiber, update : Update, methodName : ?string) : void {
+  const queue1 = ensureUpdateQueue(fiber);
+  const queue2 = fiber.alternate ? ensureUpdateQueue(fiber.alternate) : null;
+
+  // Warn if an update is scheduled from inside an updater function.
+  if (__DEV__ && typeof methodName === 'string' && (queue1.isProcessing || (queue2 && queue2.isProcessing))) {
+    if (methodName === 'setState') {
+      console.error(
+        'setState was called from inside the updater function of another' +
+        'setState. A function passed as the first argument of setState ' +
+        'should not contain any side-effects. Return a partial state object ' +
+        'instead of calling setState again. Example: ' +
+        'this.setState(function(state) { return { count: state.count + 1 }; })'
+      );
+    } else {
+      console.error(
+        `${methodName} was called from inside the updater function of ` +
+        'setState. A function passed as the first argument of setState ' +
+        'should not contain any side-effects.'
+      );
+    }
+  }
+
+  const priorityLevel = update.priorityLevel;
+
+  let queue = queue1;
+  let insertAfter1;
+  let insertBefore1;
+  let insertAfter2;
+  let insertBefore2;
+  for (let i = 0; queue && i < 2; i++) {
```

Comment:
I find the intention of this loop really confusing since it is meant to terminate after only two iterations. Seems it would be easier to just explode it but I'll try to read it as is first.